src/engine/strat_engine/devlinks.rs

```diff
@@ -0,0 +1,118 @@
+// (e.g. user shells) with the current working directory within the tree.
+pub fn setup_devlinks(engine: &Engine) -> EngineResult<()> {
+    if let Err(err) = fs::create_dir(DEV_PATH) {
+        if err.kind() != ErrorKind::AlreadyExists {
+            return Err(From::from(err));
+        }
+    }
+
+    let mut existing_dirs: HashSet<_> = fs::read_dir(DEV_PATH)?
+        .map(|dir_e| {
+                 dir_e
+                     .expect("direntry is valid")
+                     .file_name()
+                     .into_string()
+                     .expect("Unix is utf8")
+             })
+        .collect();
+
+    for pool in engine.pools() {
+        if !existing_dirs.contains(pool.name()) {
+            pool_added(pool.name());
+        }
+        existing_dirs.remove(pool.name());
+
+        let pool_path: PathBuf = vec![DEV_PATH, pool.name()].iter().collect();
+
+        let mut existing_files: HashSet<_> = fs::read_dir(pool_path)?
+            .map(|dir_e| {
+                     dir_e
+                         .expect("direntry is valid")
+                         .file_name()
+                         .into_string()
+                         .expect("Unix is utf8")
+                 })
+            .collect();
+
+        for fs in pool.filesystems() {
+            if !existing_files.contains(fs.name()) {
+                filesystem_added(pool.name(), fs.name(), &fs.devnode());
+            }
+            existing_files.remove(fs.name());
+        }
+
+        for leftover in existing_files {
+            filesystem_removed(pool.name(), &leftover)
+        }
+    }
+
+    for leftover in existing_dirs {
+        pool_removed(str::from_utf8(leftover.as_bytes()).expect("is valid utf8"))
+    }
+
+    Ok(())
+}
+
+pub fn pool_added(pool: &str) -> () {
+    let p: PathBuf = vec![DEV_PATH, pool].iter().collect();
+    fs::create_dir(&p).unwrap_or_else(|_| error!("Could not add {}", p.display()));
+}
+
+pub fn pool_removed(pool: &str) -> () {
+    let p: PathBuf = vec![DEV_PATH, pool].iter().collect();
+    fs::remove_dir_all(&p).unwrap_or_else(|_| error!("Could not remove {}", p.display()));
+}
+
+pub fn pool_renamed(old_name: &str, new_name: &str) -> () {
+    let old: PathBuf = vec![DEV_PATH, old_name].iter().collect();
+    let new: PathBuf = vec![DEV_PATH, new_name].iter().collect();
+    fs::rename(&old, &new).unwrap_or_else(|_| {
+                                              error!("Could not rename {} to {}",
+                                                     old.display(),
+                                                     new.display())
+                                          });
+}
+
+pub fn filesystem_added(pool_name: &str, fs_name: &str, devnode: &Path) -> () {
+    let p: PathBuf = vec![DEV_PATH, pool_name, fs_name].iter().collect();
+
+    // Remove existing and recreate to ensure it points to the correct devnode
+    // Failing to remove existing is ok.
```

Comment:
Why is it ok?

Reply:
It's not. Basically the entire PR was operating under a "best effort" approach, whereas now it won't.